src/tsickle.ts

```diff
@@ -867,15 +867,31 @@ class Annotator extends ClosureRewriter {
     }
   }
 
-  private visitProperty(namespace: string[], p: ts.Declaration) {
-    let name = this.propertyName(p);
+  /**
+   * @param optional If true, property is optional (e.g. written "foo?: string").
+   */
+  private visitProperty(namespace: string[], prop: ts.Declaration, optional = false) {
+    let name = this.propertyName(prop);
     if (!name) {
-      this.emit(`/* TODO: handle strange member:\n${this.escapeForComment(p.getText())}\n*/\n`);
+      this.emit(`/* TODO: handle strange member:\n${this.escapeForComment(prop.getText())}\n*/\n`);
       return;
     }
 
-    let tags = this.getJSDoc(p) || [];
-    tags.push({tagName: 'type', type: this.typeToClosure(p)});
+    let type = this.typeToClosure(prop);
+    // Interesting special case:
+    // When a property is optional, e.g.
+    //   foo?: string;
+    // Then the type of the property is string|undefined and the above
+    // translation handles it correctly.
+    // But in the special case of any, which includes undefined:
+    //   foo?: any;
+    // The type is just "any" (because that covers undefined as well) and
+    // our translation of the type is just "?".  But Closure wants it to instead
+    // be ?|undefined.
```

Comment:
I think you overlooked: "When a *property is optional*".


https://closure-compiler.appspot.com/home#code%3D%252F%252F%2520%253D%253DClosureCompiler%253D%253D%250A%252F%252F%2520%2540output_file_name%2520default.js%250A%252F%252F%2520%2540compilation_level%2520ADVANCED_OPTIMIZATIONS%250A%252F%252F%2520%253D%253D%252FClosureCompiler%253D%253D%250A%250A%252F**%2520%2540record%2520*%252F%250Afunction%2520Foo()%2520%257B%257D%250A%250A%252F**%2520%2540type%2520%257B%253F%257D%2520*%252F%250AFoo.prototype.bar%253B%250A%250A%252F**%2520%2540type%2520%257B%253F%257Cundefined%257D%2520*%252F%250AFoo.prototype.baz%253B%250A%250A%252F**%2520%2540param%2520%257B!Foo%257D%2520foo%2520*%252F%250Afunction%2520useFoo(foo)%2520%257B%2520console.log(foo)%253B%2520%257D%250A%250A%252F%252F%2520illegal%253A%250AuseFoo(%257B%257D)%253B%250AuseFoo(%257Bbaz%253A%25203%257D)%253B%250A%250A%252F%252F%2520legal%253A%250AuseFoo(%257Bbar%253A%25203%257D)%253B